Library/Homebrew/livecheck/livecheck.rb

```diff
@@ -0,0 +1,384 @@
+# frozen_string_literal: true
+
+module Homebrew
+  module Livecheck
+    module_function
+
+    GITHUB_SPECIAL_CASES = %w[
+      api.github.com
+      /latest
+      mednafen
+      camlp5
+      kotlin
+      osrm-backend
+      prometheus
+      pyenv-virtualenv
+      sysdig
+      shairport-sync
+      yuicompressor
+    ].freeze
+
+    UNSTABLE_VERSION_KEYWORDS = %w[
+      alpha
+      beta
+      bpo
+      dev
+      experimental
+      prerelease
+      preview
+      rc
+    ].freeze
+
+    def livecheck_formulae(formulae_to_check, args)
+      # Identify any non-homebrew/core taps in use for current formulae
+      non_core_taps = {}
+      formulae_to_check.each do |f|
+        next if f.tap.blank?
+        next if f.tap.name == "homebrew/core"
+
+        non_core_taps[f.tap.name] = true
+      end
+      non_core_taps = non_core_taps.keys.sort
+
+      # Load additional Strategy files from taps
+      non_core_taps.each do |tap_name|
+        tap_strategy_path = "#{Tap.fetch(tap_name).path}/livecheck/strategy"
+        Dir["#{tap_strategy_path}/*.rb"].sort.each(&method(:require)) if Dir.exist?(tap_strategy_path)
+      end
+
+      # Cache demodulized strategy names, to avoid repeating this work
```

Comment:
Would be interested in profiling this (even with `time`) to see if this actually speeds things up.

Reply:
This caching approach was added by me, so I'll explain it.

I isolated this code in a script and did some simple speed and memory profiling (using the `ruby_prof` and `memory_profiler` gems) to compare the caching approach (i.e., `@livecheck_strategy_names`) to `strategy.name.demodulize`. I started by simply doing one execution per iteration, since this happens a minimum of once per formula (unless the formula is skipped).

Using `strategy.name.demodulize` is faster until we're calling it about 50 times per `livecheck` run. At 100 executions, the caching approach is around 1.5x faster. At 1000 executions, caching is 3x faster. At 5,000 executions (around once per homebrew/core formula), caching is 3.5x faster. However, caching the demodulized strategy names is primarily about minimizing memory allocations and that's where the difference is very notable.

With caching, we allocate a small amount of memory at the start of the livecheck run and that's it. Without caching, memory is allocated each time `strategy.name.demodulize` is executed. With 5,000 iterations and one execution each, this allocates over 270x more memory than the caching approach. With two executions, it's 540x more memory. With three executions it's 810x more memory, etc.

The difference in speed hits a point where the change becomes smaller per iteration (e.g., going from 3.5x to 3.6x) but the difference in memory allocations is multiplied by the number of times `strategy.name.demodulize` is executed.

When livecheck is run with the `--debug` or `--json --verbose` flags, `@livecheck_strategy_names` is used a minimum of two times. Otherwise, `@livecheck_strategy_names` is used once per iteration.

From a speed perspective caching is beneficial above 50 executions and from a memory perspective it's beneficial above 18 executions. I do full homebrew-core runs using `--json --verbose` pretty regularly and this involves a minimum of 10,000 executions, for what it's worth.

Let me know if you need more details but hopefully this helps to clarify why it's useful.

Reply:
Fine with me, thanks for explaining 👍🏻